src/v/container/include/container/contiguous_range_map.h

```diff
@@ -0,0 +1,303 @@
+ * and limited range of integers starting at 0. The container provides a
+ * wrapper around basic chunked vector that reassembles the API of a standard
+ * map.
+ *
+ * The wrapper allows random inserts of elements to the map although this is
+ * not supported by standard vector.
+ *
+ * Underlying container is resized every time its size is not sufficient to
+ * account for emplaced key.
+ *
+ * The contiguous_range_map tolerates gap in the range of keys however number
+ * and size of gaps is proportional to performance penalty hit when incrementing
+ * or decrementing iterators.
+ *
+ * NOTE:
+ * A map can store elements for key's which range starts from value greater than
+ * 0 but it will cause significant memory loss. If we will need to store such a
+ * ranges we may introduce non type template parameter indicating offset.
+ */
+template<std::integral KeyT, typename ValueT>
+class contiguous_range_map {
+public:
+    using value_type = std::pair<const KeyT, ValueT>;
+    using key_type = KeyT;
+    using mapped_type = ValueT;
+
+private:
+    using underlying_t = chunked_vector<value_type>;
+    using unsigned_t = std::make_unsigned_t<key_type>;
+
+private:
+    template<bool Const>
+    class iter {
+    public:
+        using iterator_category = std::forward_iterator_tag;
+        using value_type =
+          typename std::conditional_t<Const, const value_type, value_type>;
+
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+        iter() = default;
+
+        /**
+         * Conversion operator allowing iterator to be converted to
+         * const_iterator, as required by the general iterator contract.
+         */
+        operator iter<true>() const { // NOLINT(hicpp-explicit-conversions)
+            iter<true> ret;
+            ret._container = _container;
+            ret._it = _it;
+            return ret;
+        }
+
+        reference operator*() const { return *_it; }
+
+        pointer operator->() const { return &(*_it); }
+
+        iter& operator++() {
+            ++_it;
+            skip_to_next_present();
+            return *this;
+        }
+
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+
+        iter operator--(int) {
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+
+        bool operator==(const iter& o) const {
+            return std::tie(_it, _container) == std::tie(o._it, o._container);
+        };
+
+        auto operator<=>(const iter& o) const {
+            return std::tie(_it, _container) <=> std::tie(o._it, o._container);
+        };
+
+    private:
+        void skip_to_next_present() {
+            while (_it != _container->_values.end()
+                   && !_container->_index_bitmap.contains(
+                     boost::numeric_cast<unsigned_t>(_it->first))) {
+                ++_it;
+            }
```

Comment:
FWIW Roaring bitmaps support iteration. Maybe that can speed this logic up?

https://github.com/RoaringBitmap/CRoaring/blob/083d39855d6ba51d3645929fcd44153a1b2f14d8/include/roaring/roaring64.h#L609-L639

Not sure if it's actually better.

Reply:
i've checked, it is not

Reply:
it seems that the bitmap is optimised for fast check if it contains a value